auto/cc/test

```diff
@@ -79,7 +79,7 @@ case $NXT_CC_NAME in
 
         # -O2 enables -fstrict-aliasing and -fstrict-overflow.
         #NXT_CFLAGS="$NXT_CFLAGS -O2"
-        #NXT_CFLAGS="$NXT_CFLAGS -Wno-strict-aliasing"
+        NXT_CFLAGS="$NXT_CFLAGS -Wno-strict-aliasing"
```

Comment:
I like this in the sense that we know Unit has Undefined Behavior in several places, and this prevents opening Hell's doors.

I don't like it in the sense that it looks like a statement that UB is allowed in this project.  I hope it's not.  From what I can see, you've been able to do quite some changes in this regard (quite surprised, I am!), so I trust UB isn't allowed.

You may also want to add `-fwrapv` (also used by the kernel) for similar reasons.


For 17288134339 ("Disable strict-aliasing in clang by default"):

```
Reviewed-by: Alejandro Colomar <alx@kernel.org>
```

Reply:
> I don't like it in the sense that it looks like a statement that UB is allowed in this project. I hope it's not

Right, we certainly don't want to generally introduce UB.

It's just unfortunate that this particular instance, that has been such a widely used idiom, and _I_ really like it, has fallen foul of compiler optimisations that I guess are allowed because the standards people didn't specify this case and so it's fair game.

I am also now armed with [this](https://blog.llvm.org/2011/05/what-every-c-programmer-should-know.html) because there is no way I can keep all this straight in my head :(.

> You may also want to add -fwrapv (also used by the kernel) for similar reasons.

Yes, good point. I can't immediately think of a reason why we shouldn't...

Reply:
Hmm, was just checking that things wrapped how I expected with -fwrapv

```c
#include <stdio.h>                                                              
#include <limits.h>                                                             
                                                                                
int main(void)                                                                  
{                                                                               
        int i = INT_MAX;                                                        
                                                                                
        printf("%d -> ", i);                                                    
        i++;                                                                    
        printf("%d\n", i);                                                      
                                                                                
        return 0;                                                               
}  

/*
 * $ ./f
 * 2147483647 -> -2147483648
 */                                                                             
```

It does, it wraps to INT_MIN. However it seems to always do that whether you compile with -fwrapv or not... maybe that's coincidence or the codes too simple and the compilers too smart...

But yes, I'll add a patch adding -fwrapv...

Reply:
I wonder if we should use, `-fno-strict-overflow` instead?

The Kernel did use -fwrapv [briefly](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=68df3755e383e6fecf2354a67b08f92f18536594) but [switched](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=a137802ee839ace40079bebde24cfb416f73208a) to `-fno-strict-overflow` a few months later...

Can't find much actual documentation on `-fno-strict-overflow` though...

Reply:
The reason the kernel stopped using it seems to be (according to the commit message that switched to `-fno-strict-overflow`) that `-fwrapv` was broken in ancient versions of GCC (4.1.x); it was supposedly fixed back in 2006, according to the GCC bugzilla <https://gcc.gnu.org/bugzilla/show_bug.cgi?id=28230>.

Curiously, that same commit message claims that `-fno-strict-overflow` is less disturbing to the compiler, but according to gcc(1), it's in fact more disturbing:

```
     -fstrict-overflow
         This option implies -fno-wrapv -fno-wrapv-pointer  and  when  negated
         implies -fwrapv -fwrapv-pointer.
```

So it implies `-fwrapv -fwrapv-pointer`.  In the case of Unit, that's a good thing, because I can point to lines of code where there are violations of pointer overflow, so indeed you *need* `-fwrapv-pointer` too.  :)

Reply:
> Hmm, was just checking that things wrapped how I expected with -fwrapv
> 
> ```c
> #include <stdio.h>                                                              
> #include <limits.h>                                                             
>                                                                                 
> int main(void)                                                                  
> {                                                                               
>         int i = INT_MAX;                                                        
>                                                                                 
>         printf("%d -> ", i);                                                    
>         i++;                                                                    
>         printf("%d\n", i);                                                      
>                                                                                 
>         return 0;                                                               
> }  
> 
> /*
>  * $ ./f
>  * 2147483647 -> -2147483648
>  */                                                                             
> ```
> 
> It does, it wraps to INT_MIN. However it seems to always do that whether you compile with -fwrapv or not... maybe that's coincidence or the codes too simple and the compilers too smart...
> 
> But yes, I'll add a patch adding -fwrapv...

Some say that Undefined Behavior waits hidden, until a cold night of February, usually with new Moon.  It doesn't reproduce when you look for it.  :)